File: stl-verify/internal/domain/entity/protocol_event.go
package entity

import (
	"encoding/json"
	"fmt"
	"time"
)

// ProtocolEvent represents a decoded protocol event stored for analytics and auditability.
type ProtocolEvent struct {
	ChainID         int
	ProtocolID      int64
	BlockNumber     int64
	BlockVersion    int
	TxHash          []byte
	LogIndex        int
	ContractAddress []byte
	EventName       string
	EventData       json.RawMessage
	CreatedAt       time.Time // block timestamp — deterministic for hypertable dedup
}

// NewProtocolEvent creates a new ProtocolEvent with validation.
func NewProtocolEvent(chainID int, protocolID, blockNumber int64, blockVersion int, txHash []byte, logIndex int, contractAddress []byte, eventName string, eventData json.RawMessage, createdAt time.Time) (*ProtocolEvent, error) {
	e := &ProtocolEvent{
		ChainID:         chainID,
		ProtocolID:      protocolID,
		BlockNumber:     blockNumber,
		BlockVersion:    blockVersion,
		TxHash:          txHash,
		LogIndex:        logIndex,
		ContractAddress: contractAddress,
		EventName:       eventName,
		EventData:       eventData,
		CreatedAt:       createdAt,
	}
	if err := e.Validate(); err != nil {
		return nil, fmt.Errorf("NewProtocolEvent: %w", err)
	}
	return e, nil
}

func (e *ProtocolEvent) Validate() error {
	if e.CreatedAt.IsZero() {
		return fmt.Errorf("createdAt must be set explicitly (block timestamp)")
	}
	if e.ChainID <= 0 {
		return fmt.Errorf("chainID must be positive, got %d", e.ChainID)
	}
	if e.ProtocolID <= 0 {
		return fmt.Errorf("protocolID must be positive, got %d", e.ProtocolID)
	}
	// Non-negative (allows block 0 / genesis) — unlike position entities which require
	// positive block numbers, events can occur in the genesis block (e.g. contract deploys).
	if e.BlockNumber < 0 {
		return fmt.Errorf("blockNumber must be non-negative, got %d", e.BlockNumber)
	}
	if e.BlockVersion < 0 {
		return fmt.Errorf("blockVersion must be non-negative, got %d", e.BlockVersion)
	}
	if len(e.TxHash) == 0 {
		return fmt.Errorf("txHash must not be empty")
	}
	if e.LogIndex < 0 {
		return fmt.Errorf("logIndex must be non-negative, got %d", e.LogIndex)
	}
	if len(e.ContractAddress) == 0 {
		return fmt.Errorf("contractAddress must not be empty")
	}
	if e.EventName == "" {
		return fmt.Errorf("eventName must not be empty")
	}
	if len(e.EventData) == 0 {
		return fmt.Errorf("eventData must not be empty")
	}
	if !json.Valid(e.EventData) {
		return fmt.Errorf("eventData must be valid JSON")
	}
	return nil
}

File: stl-verify/python/app/domain/entities/protocol_event.py
import re
from dataclasses import dataclass
from datetime import datetime
from typing import Any

_TX_HASH_RE = re.compile(r"^0x[0-9a-fA-F]{64}$")
_ETH_ADDRESS_RE = re.compile(r"^0x[0-9a-fA-F]{40}$")


@dataclass(frozen=True)
class ProtocolEvent:
    """Protocol event detail with transaction context."""

    tx_hash: str
    log_index: int
    chain_id: int
    block_number: int
    block_version: int
    protocol_name: str
    event_name: str
    contract_address: str
    event_data: dict[str, Any] | None
    created_at: datetime

    def __post_init__(self) -> None:
        if not isinstance(self.tx_hash, str) or not _TX_HASH_RE.match(self.tx_hash):
            raise ValueError(f"tx_hash must be a valid transaction hash (0x + 64 hex chars), got {self.tx_hash!r}")
        if self.log_index < 0:
            raise ValueError(f"log_index must be non-negative, got {self.log_index}")
        if self.chain_id <= 0:
            raise ValueError(f"chain_id must be positive, got {self.chain_id}")
        if self.block_number <= 0:
            raise ValueError(f"block_number must be positive, got {self.block_number}")
        if self.block_version < 0:
            raise ValueError(f"block_version must be non-negative, got {self.block_version}")
        if not self.protocol_name or not self.protocol_name.strip():
            raise ValueError("protocol_name must be non-empty")
        if not self.event_name or not self.event_name.strip():
            raise ValueError("event_name must be non-empty")
        if not isinstance(self.contract_address, str) or not _ETH_ADDRESS_RE.match(self.contract_address):
            raise ValueError(
                f"contract_address must be a valid Ethereum address (0x + 40 hex chars), got {self.contract_address!r}"
            )
